sycl/doc/dev/RuntimeTestingGuidelines.md

```diff
@@ -0,0 +1,123 @@
+> on low-level runtime call result, use PI Mock to simulate faulty call.
+
+### Device Image Mock
+
+SYCL Runtime discoveres user-defined kernels by looking up kernel names in
+registered device images. These images contain all information about SPIR-V or
+binary device code, including kernel names and some metadata passed from the
+compiler frontend.
+
+`helpers/PIImage.hpp` header provides routines to insert a fake device image
+into Program Manager, and emulate presence of particular kernels.
+
+In the default compiler flow, Clang generates a so-called integration header,
+which contains information about available kernels. When compiling unit tests,
+the device compiler is not used, and developers must provide specializations
+for `KernelInfo` class template. See `kernel_desc.hpp` for more info.
+
+### Other helpers
+
+#### Mock Scheduler
+
+For the purpose of testing SYCL Scheduler without submitting kernels to the
+queue, there're some mock classes, that use Google Mock framework.
+See `syc/unittests/scheduler/SchedulerTestUtils.hpp` for more info, and refer
+to other scheduler tests as an example.
+
+#### Testing environment variables
+
+To facilitate testing environment variables and config controls, there's a
+`ScopedEnvVar` helper class, which is a RAII-style wrapper around `setenv` and
+`getenv`. See `helpers/ScopedEnvVar.hpp` for more info.
+
+### Changing ABI
+
+Whenever a new data type is added in the pull request as part of SYCL library
+ABI (i.e. the data structure crosses library boundaries), such pull requests
+must also provide a test, that would cover the layout of that data type.
+
+New APIs are also covered by ABI testing. The testing tool will suggest steps
+if a new API is added to SYCL runtime library.
+
+Refer to [ABI Policy Guide](https://intel.github.io/llvm-docs/ABIPolicyGuide.html)
+for more info.
+
+## SYCL Device Headers
+
+SYCL headers often use "magical" compiler builtins, that are later lowered to
+device-specific instruction sequences. Whenever Pull Request adds or changes
+uses of those builtins, it must also provide a LIT test, that would check
+the correctness of compiler frontend output in the form of LLVM IR (i.e.
+check calls to particular intrinsics exist, function mangling is correct, etc).
+
+Find examples of such tests in `sycl/test/check_device_code` directory.
+
+Refer to [FileCheck](https://llvm.org/docs/CommandGuide/FileCheck.html) and
+[LIT](https://llvm.org/docs/CommandGuide/lit.html) tools documentation for
+more info.
+
+### Changing ABI
+
+Whenever a new data structure is added as part of ABI (either between device and
+host, or between two device images), a test for data layout must be added,
+same as for host runtime. If the data structure can be used both on host and
+device sides of user code, both tests must be present.
+
+Refer to [ABI Policy Guide](https://intel.github.io/llvm-docs/ABIPolicyGuide.html)
+for more info.
+
+## End-to-end tests
```

Comment:
Do we want to check expected warnings/errors in  https://github.com/intel/llvm/tree/sycl/sycl/test/warnings?